### PR TITLE
Use builtin overflow functions of Clang and GCC (WIP, RFC)

### PR DESCRIPTION
**Pull request by def-**
This only works with recent Clang and GCC 5.0.

The resulting code is about 3 times faster than the current one for overflow checks on my system. I tested this with `nim c --cc:clang -d:release --checks:on c file`:

```nimrod
var x = 10000000000.int
var y = 0
while true:
  y += x
echo y
```

Currently it doesn't work because __has_builtin can't be defined.

This is totally ugly. But we can't reliably detect whether we can use
the builtins from Nim, especially with cross-compiling where the user
may be using an older compiler version. Switching this on/off manually
with a define seems weird as well, this should work automatically.

### CODE REVIEW

**Comment by ReneSac** [comment]
Those bare emits will break the js backend, right? Nim needs a standard and clean method to do `#ifdef` to interface with C. So among other things we will not need to use 'gettimeofday()' for the GC on linux.

**Comment by def-** [comment]
Ideally I'd want this behind a `when clang > XX or gcc >= 5.0:` instead.

**Comment by Araq** [comment]
@ReneSac That code is not active for the JS backend anyway.

File: lib/system/arithm.nim [hunk 1]
```diff
@@ -8,6 +8,75 @@
...
+#endif""".}
+
+# Builtin compiler functions for improved performance
+
+proc checkFunction(name: string): string =
+  "((__has_builtin(__builtin_" & name & "_overflow)) || __GNUC__ >= 5)"
+
+# TODO: This is totally ugly. But we can't reliably detect this from Nim,
+# especially with cross-compiling where the user may be using an older compiler
+# version. Switching this on/off manually with a define seems weird as well.
```
Comment (Araq) [inline lib/system/arithm.nim]: I think this should be done by a define. Once GCC 5.0 becomes widespread, we can enable it as a default in the config, for now it should be opt-in.

File: lib/system/arithm.nim [hunk 2]
```diff
@@ -8,6 +8,75 @@
 #
 
 
+# Only clang has __has_builtin (so far)
+#
+# TODO: This is emitted at the wrong position so we don't actually have an
```
Comment (Araq) [inline lib/system/arithm.nim]: Yeah add it to nimbase.h instead.

**Comment by def-** [comment]
Alright, moved behind much simpler `-d:builtinOverflow`:

> Use -d:builtinOverflow for builtin overflow checks of recent Clang and GCC >= 5.0.
> 
> Maybe it's better that it's optional after all. GCC's code with builtin
> overflow checks seems to be a bit slower actually, while Clang's is 3
> times faster.

**Comment by Varriount** [comment]
@Araq Can this be merged?

**Pull request merged**